### ISSUE TITLE
Enforce DISTRIBUTED BY check for PRIMARY KEY/CONSTRAINT UNIQUE in ALTER TABLE

### ISSUE THREAD

**Issue by GodenYao** [greenplum-db/gpdb#394]
In: https://github.com/greenplum-db/gpdb/blob/master/src/backend/parser/analyze.c

We added check while creating tables "DISTRIBUTED BY" along with PRIMARY KEY or CONSTRAINT UNIQUE to make sure distribution key is equal or left-subset of the primary key.

However, user can bypass this check by ALTER TABLE to change the distribution key later.
Given it's crucial to have the correct distribution key to achieve maximum parallelism, we need to enforce that check in ALTER TABLE as well.

**Comment by ivannovick** [comment]
@vraghavan78 @hlinnaka is this a valid issue?

**Comment by hlinnaka** [comment]
Yeah, seems we should be consistent, and perform similar checks in CREATE TABLE and ALTER TABLE.

Why does that check in the first place, BTW? You mention maximum parallelism, but I wonder if it's actually because we have no way to enforce a unique constraint across segments?

**Comment by hlinnaka** [comment]
This report didn't actually include any details on what exactly is the sequence of commands that we're not enforcing correctly. Looks like I formed an opinion back in Jan 2017, but now that I look at this, I'm not sure what this was about. Has it been fixed, since?

**Comment by hlinnaka** [comment]
I was looking at this in the context of https://groups.google.com/a/greenplum.org/forum/#!topic/gpdb-dev/OJzz8I6WuVI and PR #5902. Does PR #5902 fix this issue?

**Comment by hlinnaka** [comment]
Hmm. The HINT in the error message is not quite right, though. It doesn't have to be a left-subset, any subset will do. I'll go fix that.
